### PR TITLE
[1251] Implement state machine for user, add transitioned state

### PR DESCRIPTION
**Pull request by tvararu**
### Context

This is more or less a fork of https://github.com/DFE-Digital/manage-courses-backend/pull/274.

### Changes proposed in this pull request

- Implements the state machine pattern on the users model, with initial states around the users `transitioned`

### CODE REVIEW

File: db/data/20190418114011_initialise_state_for_users.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    User.find_each do |user|
+      user.save!
+    end
+  end
+
+  def down
+    User.find_each do |user|
+      next unless user.new?
+
+      user.update(aasm_state: '')
```
Comment (dankmitchell) [inline db/data/20190418114011_initialise_state_for_users.rb]: Not sure this is right. There will always be an initial state set in the model i.e. `state :new, initial: true`

File: db/data/20190418114011_initialise_state_for_users.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+class InitialiseStateForUsers < ActiveRecord::Migration[5.2]
+  def up
+    User.find_each do |user|
```
Comment (dankmitchell) [inline db/data/20190418114011_initialise_state_for_users.rb]: Could just be a oneliner `User.find_each(&:save)`

Comment (tvararu) [reply db/data/20190418114011_initialise_state_for_users.rb]: Done already by rubocop

File: db/data/20190418114011_initialise_state_for_users.rb (hunk 1 above)
Comment (tvararu) [reply db/data/20190418114011_initialise_state_for_users.rb]: If we do that then it's not a full reverse-migration; the initial database state is empty string. 🤔

File: app/models/user.rb [hunk 3]
```diff
@@ -12,13 +12,25 @@
...
+
   has_and_belongs_to_many :organisations
   has_many :providers, through: :organisations
 
   validates :email, presence: true
 
   audited
+
+  aasm column: 'state' do
+    state :new, initial: true
```
Comment (heidar) [inline app/models/user.rb]: Would it be better to name these states based on what we use them for? I think we only use them for terms and conditions?

File: db/data/20190418114011_initialise_state_for_users.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class InitialiseStateForUsers < ActiveRecord::Migration[5.2]
+  def up
+    User.find_each(&:save!)
```
Comment (heidar) [inline db/data/20190418114011_initialise_state_for_users.rb]: It'd be more explicit to make this set `state` to `User.aasm.initial_state`. Not a dealbreaker though. :)

File: app/models/user.rb (hunk 3 above)
Comment (tvararu) [reply app/models/user.rb]: Discussed offline, it's going to be used for Ts and Cs as well but it's ok.

File: db/data/20190418114011_initialise_state_for_users.rb (hunk 4 above)
Comment (tvararu) [reply db/data/20190418114011_initialise_state_for_users.rb]: Yes, agree

**Pull request merged**